Regression: Queries with inverse property path expression throw AssertionError in Var because parent is already set

### Current Behavior

### Current Behavior

When parsing the following query with RDF4J 4.1.2 (worked with 4.1.0 and 4.0.0) I get an AssertionError:

```
PREFIX foaf: <http://xmlns.com/foaf/0.1/>
SELECT *
{
  ?x foaf:knows/^foaf:knows ?y .  
  FILTER(?x != ?y)
}
```

This might be related to recent optimizations (also the follow-ups for #4157) and is likely missing one of the Expr.clone() calls or so?

Stacktrace:
```
java.lang.AssertionError: Unexpected assertion
	at com.metaphacts.rdf4j.RDF4JTest.testVar(RDF4JTest.java:20)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.lang.AssertionError
	at org.eclipse.rdf4j.query.algebra.Var.setParentNode(Var.java:112)
	at org.eclipse.rdf4j.query.algebra.StatementPattern.<init>(StatementPattern.java:101)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:1486)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:237)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTPathElt.jjtAccept(ASTPathElt.java:32)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:1411)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:237)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTPathSequence.jjtAccept(ASTPathSequence.java:30)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:1358)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:237)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTPathAlternative.jjtAccept(ASTPathAlternative.java:28)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:1663)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTPropertyListPath.jjtAccept(ASTPropertyListPath.java:28)
	at org.eclipse.rdf4j.query.parser.sparql.ast.SimpleNode.childrenAccept(SimpleNode.java:102)
	at org.eclipse.rdf4j.query.parser.sparql.AbstractASTVisitor.visit(AbstractASTVisitor.java:846)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:2633)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTTriplesSameSubjectPath.jjtAccept(ASTTriplesSameSubjectPath.java:28)
	at org.eclipse.rdf4j.query.parser.sparql.ast.SimpleNode.childrenAccept(SimpleNode.java:102)
	at org.eclipse.rdf4j.query.parser.sparql.AbstractASTVisitor.visit(AbstractASTVisitor.java:381)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTBasicGraphPattern.jjtAccept(ASTBasicGraphPattern.java:27)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:1197)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:237)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTGraphPatternGroup.jjtAccept(ASTGraphPatternGroup.java:27)
	at org.eclipse.rdf4j.query.parser.sparql.ast.SimpleNode.childrenAccept(SimpleNode.java:102)
	at org.eclipse.rdf4j.query.parser.sparql.AbstractASTVisitor.visit(AbstractASTVisitor.java:880)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTWhereClause.jjtAccept(ASTWhereClause.java:27)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:355)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:237)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTSelectQuery.jjtAccept(ASTSelectQuery.java:27)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:346)
	at org.eclipse.rdf4j.query.parser.sparql.TupleExprBuilder.visit(TupleExprBuilder.java:237)
	at org.eclipse.rdf4j.query.parser.sparql.ast.ASTQueryContainer.jjtAccept(ASTQueryContainer.java:29)
	at org.eclipse.rdf4j.query.parser.sparql.SPARQLParser.buildQueryModel(SPARQLParser.java:246)
	at org.eclipse.rdf4j.query.parser.sparql.SPARQLParser.parseQuery(SPARQLParser.java:206)
	at com.metaphacts.rdf4j.RDF4JTest.testVar(RDF4JTest.java:17)
	... 26 more

```


### Expected Behavior

The query can be parsed without errors.

### Steps To Reproduce

This unit test reproduces the behavior:

Removing the inverse operator in the property path (`foaf:knows/^foaf:knows` -> `foaf:knows/foaf:knows`) makes it pass, so the issue must be somewhere in there.
```
package com.metaphacts.rdf4j;

import org.eclipse.rdf4j.query.parser.sparql.SPARQLParser;
import org.junit.Test;

public class RDF4JTest {
	@Test
	public void testVar() throws Exception {
		String queryString = "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n"
				+ "  SELECT *\n"
				+ "  {\n"
				+ "    ?x foaf:knows/^foaf:knows ?y .  \n"
				+ "    FILTER(?x != ?y)\n"
				+ "  }\n";
		SPARQLParser parser = new SPARQLParser();
		try {
			parser.parseQuery(queryString, null);
		}
		catch (AssertionError e) {
			throw new AssertionError("Unexpected assertion", e);
		}
	}
}
```

### Version

4.1.2

### Are you interested in contributing a solution yourself?

_No response_

### Anything else?

_No response_

--- a/core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/TupleExprBuilder.java
+++ b/core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/TupleExprBuilder.java
@@ -275,11 +275,13 @@ public class TupleExprBuilder extends AbstractASTVisitor {
 	 */
 	protected Var mapValueExprToVar(Object valueExpr) {
 		if (valueExpr instanceof Var) {
-			return (Var) valueExpr;
+			// Return a clone to prevent the Var object from being used in more than one place.
+			return ((Var) valueExpr).clone();
 		} else if (valueExpr instanceof ValueConstant) {
 			return TupleExprs.createConstVar(((ValueConstant) valueExpr).getValue());
 		} else if (valueExpr instanceof TripleRef) {
-			return ((TripleRef) valueExpr).getExprVar();
+			// Return a clone to prevent the Var object from being used in more than one place.
+			return ((TripleRef) valueExpr).getExprVar().clone();
 		} else if (valueExpr == null) {
 			throw new IllegalArgumentException("valueExpr is null");
 		} else {
@@ -1480,9 +1482,9 @@ public class TupleExprBuilder extends AbstractASTVisitor {
 			}
 			pathElementExpression = createTupleExprForNegatedPropertySets(psElems, pathSequenceContext);
 		} else {
-			final ValueExpr pred = (ValueExpr) pathElement.jjtGetChild(0).jjtAccept(this, pathSequenceContext);
-			final Var predVar = mapValueExprToVar(pred);
-			pathElementExpression = new StatementPattern(pathSequenceContext.scope, startVar.clone(), predVar, endVar,
+			Var predVar = mapValueExprToVar(pathElement.jjtGetChild(0).jjtAccept(this, pathSequenceContext));
+			pathElementExpression = new StatementPattern(pathSequenceContext.scope, startVar.clone(), predVar,
+					endVar.clone(),
 					pathSequenceContext.contextVar != null ? pathSequenceContext.contextVar.clone() : null);
 		}
 
@@ -1719,8 +1721,8 @@ public class TupleExprBuilder extends AbstractASTVisitor {
 		for (int i = 0; i < childCount; i++) {
 			ValueExpr childValue = (ValueExpr) node.jjtGetChild(i).jjtAccept(this, null);
 
-			Var childVar = mapValueExprToVar(childValue);
-			graphPattern.addRequiredSP(listVar.clone(), TupleExprs.createConstVar(RDF.FIRST), childVar.clone());
+			graphPattern.addRequiredSP(listVar.clone(), TupleExprs.createConstVar(RDF.FIRST),
+					mapValueExprToVar(childValue));
 
 			Var nextListVar;
 			if (i == childCount - 1) {
@@ -2619,9 +2621,9 @@ public class TupleExprBuilder extends AbstractASTVisitor {
 	@Override
 	public TupleExpr visit(ASTTripleRef node, Object data) throws VisitorException {
 		TripleRef ret = new TripleRef();
-		ret.setSubjectVar(mapValueExprToVar(node.getSubj().jjtAccept(this, ret)).clone());
-		ret.setPredicateVar(mapValueExprToVar(node.getPred().jjtAccept(this, ret)).clone());
-		ret.setObjectVar(mapValueExprToVar(node.getObj().jjtAccept(this, ret)).clone());
+		ret.setSubjectVar(mapValueExprToVar(node.getSubj().jjtAccept(this, ret)));
+		ret.setPredicateVar(mapValueExprToVar(node.getPred().jjtAccept(this, ret)));
+		ret.setObjectVar(mapValueExprToVar(node.getObj().jjtAccept(this, ret)));
 		ret.setExprVar(createAnonVar());
 		graphPattern.addRequiredTE(ret);
 

--- a/testsuites/repository/src/main/java/org/eclipse/rdf4j/testsuite/repository/RDFStarSupportTest.java
+++ b/testsuites/repository/src/main/java/org/eclipse/rdf4j/testsuite/repository/RDFStarSupportTest.java
@@ -11,12 +11,11 @@
 package org.eclipse.rdf4j.testsuite.repository;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
+import java.math.BigInteger;
 import java.util.List;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
 
 import org.eclipse.rdf4j.model.BNode;
 import org.eclipse.rdf4j.model.IRI;
@@ -24,6 +23,7 @@ import org.eclipse.rdf4j.model.Literal;
 import org.eclipse.rdf4j.model.Statement;
 import org.eclipse.rdf4j.model.Triple;
 import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.util.Values;
 import org.eclipse.rdf4j.model.vocabulary.FOAF;
 import org.eclipse.rdf4j.model.vocabulary.RDF;
 import org.eclipse.rdf4j.query.BindingSet;
@@ -31,6 +31,7 @@ import org.eclipse.rdf4j.query.GraphQuery;
 import org.eclipse.rdf4j.query.GraphQueryResult;
 import org.eclipse.rdf4j.query.QueryResults;
 import org.eclipse.rdf4j.query.TupleQuery;
+import org.eclipse.rdf4j.query.TupleQueryResult;
 import org.eclipse.rdf4j.repository.Repository;
 import org.eclipse.rdf4j.repository.RepositoryConnection;
 import org.junit.jupiter.api.AfterEach;
@@ -43,33 +44,24 @@ import org.junit.jupiter.api.Timeout;
  * Test cases for RDF-star support in the Repository.
  *
  * @author Jeen Broekstra
- *
  */
 @Timeout(value = 10, unit = TimeUnit.MINUTES)
 public abstract class RDFStarSupportTest {
 
 	private Repository testRepository;
 
-	private RepositoryConnection testCon;
+	protected RepositoryConnection testCon;
 
 	private ValueFactory vf;
 
 	private BNode bob;
-
 	private BNode alice;
-
 	private BNode alexander;
-
 	private Literal nameAlice;
-
 	private Literal nameBob;
-
 	private Literal mboxAlice;
-
 	private Literal mboxBob;
-
 	private IRI context1;
-
 	private IRI context2;
 
 	@BeforeEach
@@ -107,7 +99,7 @@ public abstract class RDFStarSupportTest {
 	}
 
 	@Test
-	public void testAddRDFStarSubject() throws Exception {
+	public void testAddRDFStarSubject() {
 		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
 
 		testCon.add(rdfStarTriple, RDF.TYPE, RDF.ALT);
@@ -116,7 +108,7 @@ public abstract class RDFStarSupportTest {
 	}
 
 	@Test
-	public void testAddRDFStarObject() throws Exception {
+	public void testAddRDFStarObject() {
 		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
 
 		testCon.add(RDF.ALT, RDF.TYPE, rdfStarTriple);
@@ -125,12 +117,12 @@ public abstract class RDFStarSupportTest {
 	}
 
 	@Test
-	public void testAddRDFStarContext() throws Exception {
+	public void testAddRDFStarContext() {
 		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
 
 		try {
 			testCon.add(RDF.ALT, RDF.TYPE, RDF.ALT, rdfStarTriple);
-			fail("RDF-star triple value should not be allowed by store as context identifier");
+			Assertions.fail("RDF-star triple value should not be allowed by store as context identifier");
 		} catch (UnsupportedOperationException e) {
 			// fall through, expected behavior
 			testCon.rollback();
@@ -144,7 +136,8 @@ public abstract class RDFStarSupportTest {
 
 		testCon.add(rdfStarTriple, RDF.TYPE, RDF.ALT);
 
-		String query = "PREFIX foaf: <" + FOAF.NAMESPACE + ">\nSELECT DISTINCT * WHERE { <<?s foaf:name ?o>> ?b ?c. }";
+		String query = "PREFIX foaf: <" + FOAF.NAMESPACE + ">\n" +
+				"SELECT DISTINCT * WHERE { <<?s foaf:name ?o>> ?b ?c. }";
 
 		List<BindingSet> result = QueryResults.asList(testCon.prepareTupleQuery(query).evaluate());
 		assertThat(result).hasSize(1);
@@ -162,16 +155,16 @@ public abstract class RDFStarSupportTest {
 		Triple rdfStarTriple = vf.createTriple(bob, FOAF.NAME, nameBob);
 		testCon.add(rdfStarTriple, RDF.TYPE, RDF.ALT);
 
-		String update = "PREFIX foaf: <" + FOAF.NAMESPACE
-				+ ">\n INSERT { ?s foaf:age 23 } WHERE { <<?s foaf:name ?o>> ?b ?c .}";
+		String update = "PREFIX foaf: <" + FOAF.NAMESPACE + ">\n" +
+				"INSERT { ?s foaf:age 23 } WHERE { <<?s foaf:name ?o>> ?b ?c .}";
 
 		testCon.prepareUpdate(update).execute();
 
-		assertThat(testCon.hasStatement(bob, FOAF.AGE, vf.createLiteral(23), false));
+		Assertions.assertTrue(testCon.hasStatement(bob, FOAF.AGE, vf.createLiteral(BigInteger.valueOf(23)), false));
 	}
 
 	@Test
-	public void testRdfStarAddAndRetrieveSparql() throws InterruptedException {
+	public void testRdfStarAddAndRetrieveSparql() {
 
 		Triple insertedTriple = vf.createTriple(RDF.SUBJECT, RDF.PREDICATE, RDF.OBJECT);
 
@@ -183,13 +176,13 @@ public abstract class RDFStarSupportTest {
 		TupleQuery query = testCon.prepareTupleQuery(
 				"SELECT * WHERE { << <http://www.w3.org/1999/02/22-rdf-syntax-ns#subject> <http://www.w3.org/1999/02/22-rdf-syntax-ns#predicate> <http://www.w3.org/1999/02/22-rdf-syntax-ns#object> >> ?a ?b}");
 
-		assertTrue(testCon.prepareBooleanQuery("ASK { ?t a 'I am a triple ;-D'}").evaluate());
-		assertEquals(1, query.evaluate().stream().count());
+		Assertions.assertTrue(testCon.prepareBooleanQuery("ASK { ?t a 'I am a triple ;-D'}").evaluate());
+		Assertions.assertEquals(1, getCount(query));
 		testCon.commit();
 	}
 
 	@Test
-	public void testRdfStarAddAndRetrieveSparqlSeparateTransaction() throws InterruptedException {
+	public void testRdfStarAddAndRetrieveSparqlSeparateTransaction() {
 
 		Triple insertedTriple = vf.createTriple(RDF.SUBJECT, RDF.PREDICATE, RDF.OBJECT);
 		Literal literal = vf.createLiteral("I am a triple ;-D");
@@ -198,18 +191,22 @@ public abstract class RDFStarSupportTest {
 		testCon.add(insertedTriple, RDF.TYPE, literal);
 		testCon.commit();
 		testCon.begin();
-		assertTrue(testCon.prepareBooleanQuery("ASK { ?t a 'I am a triple ;-D'}").evaluate());
-		assertEquals(1, testCon.prepareTupleQuery(
-				"SELECT * WHERE { << <http://www.w3.org/1999/02/22-rdf-syntax-ns#subject> <http://www.w3.org/1999/02/22-rdf-syntax-ns#predicate> <http://www.w3.org/1999/02/22-rdf-syntax-ns#object> >> ?a ?b}")
-				.evaluate()
-				.stream()
-				.count());
+		Assertions.assertTrue(testCon.prepareBooleanQuery("ASK { ?t a 'I am a triple ;-D'}").evaluate());
+		TupleQuery tupleQuery = testCon.prepareTupleQuery(
+				"SELECT * WHERE { << <http://www.w3.org/1999/02/22-rdf-syntax-ns#subject> <http://www.w3.org/1999/02/22-rdf-syntax-ns#predicate> <http://www.w3.org/1999/02/22-rdf-syntax-ns#object> >> ?a ?b}");
+		Assertions.assertEquals(1, getCount(tupleQuery));
 		testCon.commit();
 
 	}
 
+	private static long getCount(TupleQuery tupleQuery) {
+		try (TupleQueryResult evaluate = tupleQuery.evaluate()) {
+			return evaluate.stream().count();
+		}
+	}
+
 	@Test
-	public void testRdfStarAddAndRetrieve() throws InterruptedException {
+	public void testRdfStarAddAndRetrieve() {
 
 		Triple insertedTriple = vf.createTriple(RDF.SUBJECT, RDF.PREDICATE, RDF.OBJECT);
 		Triple copyOfInsertedTriple = vf.createTriple(RDF.SUBJECT, RDF.PREDICATE, RDF.OBJECT);
@@ -218,8 +215,8 @@ public abstract class RDFStarSupportTest {
 
 		testCon.add(insertedTriple, RDF.TYPE, literal);
 
-		assertEquals(1, testCon.getStatements(null, RDF.TYPE, literal, false).stream().count());
-		assertEquals(1, testCon.getStatements(copyOfInsertedTriple, null, null, false).stream().count());
+		Assertions.assertEquals(1, testCon.getStatements(null, RDF.TYPE, literal, false).stream().count());
+		Assertions.assertEquals(1, testCon.getStatements(copyOfInsertedTriple, null, null, false).stream().count());
 		testCon.commit();
 
 	}
@@ -271,6 +268,18 @@ public abstract class RDFStarSupportTest {
 				List<Statement> statements = QueryResults.asList(result);
 				Assertions.assertEquals(29, statements.size());
 			}
+		}
+	}
+
+	@Test
+	public void testSparqlStarInObjectPosition() {
+		testCon.add(Values.bnode(), FOAF.KNOWS, Values.triple(Values.bnode(), FOAF.NAME, Values.literal("John Doe")));
+		TupleQuery tupleQuery = testCon.prepareTupleQuery(
+				"PREFIX foaf: <" + FOAF.NAMESPACE + ">\n" +
+						"SELECT * WHERE { ?a ?b <<?s foaf:name ?o>>. }");
+		try (TupleQueryResult evaluate = tupleQuery.evaluate()) {
+			List<BindingSet> collect = evaluate.stream().collect(Collectors.toList());
+			Assertions.assertEquals(1, collect.size());
 		}
 	}
 

--- a/testsuites/sparql/src/main/java/org/eclipse/rdf4j/testsuite/sparql/tests/PropertyPathTest.java
+++ b/testsuites/sparql/src/main/java/org/eclipse/rdf4j/testsuite/sparql/tests/PropertyPathTest.java
@@ -21,12 +21,17 @@ import static org.junit.Assert.fail;
 
 import java.io.StringReader;
 import java.util.List;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import org.eclipse.rdf4j.model.BNode;
 import org.eclipse.rdf4j.model.IRI;
 import org.eclipse.rdf4j.model.Literal;
 import org.eclipse.rdf4j.model.Value;
 import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.util.Values;
+import org.eclipse.rdf4j.model.vocabulary.FOAF;
+import org.eclipse.rdf4j.model.vocabulary.SKOS;
 import org.eclipse.rdf4j.query.Binding;
 import org.eclipse.rdf4j.query.BindingSet;
 import org.eclipse.rdf4j.query.QueryEvaluationException;
@@ -226,6 +231,41 @@ public class PropertyPathTest extends AbstractComplianceTest {
 
 		List<BindingSet> result = QueryResults.asList(tq.evaluate());
 		assertThat(result).hasSize(4);
+	}
+
+	@Test
+	public void testComplexPath() {
+		conn.add(Values.bnode(), SKOS.BROADER, Values.bnode());
+		conn.add(Values.bnode(), SKOS.TOP_CONCEPT_OF, Values.bnode());
+
+		TupleQuery tupleQuery = conn.prepareTupleQuery("PREFIX skos:<http://www.w3.org/2004/02/skos/core#> \r\n" +
+				" SELECT *  " +
+				" WHERE {\r\n" +
+				"   ?s (skos:broader|^skos:narrower|skos:topConceptOf|^skos:hasTopConcept) ?o.\r\n" +
+				" }");
+		try (TupleQueryResult evaluate = tupleQuery.evaluate()) {
+			List<BindingSet> collect = evaluate.stream().collect(Collectors.toList());
+			assertEquals(2, collect.size());
+		}
+	}
+
+	@Test
+	public void testInversePath() {
+		BNode bnode1 = Values.bnode("bnode1");
+
+		conn.add(Values.bnode(), FOAF.KNOWS, bnode1);
+		conn.add(Values.bnode(), FOAF.KNOWS, bnode1);
+
+		TupleQuery tupleQuery = conn.prepareTupleQuery("PREFIX foaf: <" + FOAF.NAMESPACE + ">\n" +
+				"SELECT * WHERE {\n" +
+				"  ?x foaf:knows/^foaf:knows ?y . \n" +
+				"  FILTER(?x != ?y)\n" +
+				"}");
+
+		try (TupleQueryResult evaluate = tupleQuery.evaluate()) {
+			List<BindingSet> collect = evaluate.stream().collect(Collectors.toList());
+			assertEquals(2, collect.size());
+		}
 	}
 
 	private boolean containsSolution(List<BindingSet> result, Binding... solution) {

